syncs/github-discussions/main.ts

```diff
@@ -0,0 +1,141 @@
+//  _ __ ___   ___ _ __ __ _  ___ ___| |_ __ _| |_
+// | '_ ` _ \ / _ | '__/ _` |/ _ / __| __/ _` | __|
+// | | | | | |  __| | | (_| |  __\__ | || (_| | |_
+// |_| |_| |_|\___|_|  \__, |\___|___/\__\__,_|\__|
+//                     |___/
+//
+// This syncer uses the GitHub API to sync discussions for the given repository.
+//
+// @author: Patrick DeVivo (patrick@mergestat.com)
+
+import { Octokit } from "https://esm.sh/v124/octokit@2.0.14";
+import { paginateGraphql } from "https://cdn.jsdelivr.net/npm/@octokit/plugin-paginate-graphql@2.0.1/+esm";
+import { throttling } from "https://esm.sh/@octokit/plugin-throttling";
+import { Client } from "https://deno.land/x/postgres@v0.17.0/mod.ts";
+
+const discussionsQuery = await Deno.readTextFile("./query.gql");
+const params = JSON.parse(Deno.env.get("MERGESTAT_PARAMS") || "{}");
+const repoID = Deno.env.get("MERGESTAT_REPO_ID")
+const repoURL = new URL(Deno.env.get("MERGESTAT_REPO_URL") || "");
+const owner = repoURL.pathname.split("/")[1];
+const repo = repoURL.pathname.split("/")[2];
+
+const OctokitWithGrapQLPagination = Octokit.plugin(paginateGraphql, throttling);
+const octokit = new OctokitWithGrapQLPagination({
+    auth: Deno.env.get("MERGESTAT_AUTH_TOKEN"),
+    throttle: {
+        onRateLimit: (retryAfter, options, octokit, retryCount) => {
+          octokit.log.warn(
+            `Request quota exhausted for request ${options.method} ${options.url}`
+          );
+    
+          if (retryCount < 1) {
+            // only retries once
+            octokit.log.info(`Retrying after ${retryAfter} seconds!`);
+            return true;
+          }
+        },
+        onSecondaryRateLimit: (retryAfter, options, octokit) => {
+          // does not retry, only logs a warning
+          octokit.log.warn(
+            `SecondaryRateLimit detected for request ${options.method} ${options.url}`
+          );
+        },
+      },
+});
+
+const discussionsBuffer = [];
+const perPage = params.perPage || 100;
+
+const iterator = octokit.graphql.paginate.iterator(discussionsQuery, {
+    owner, repo, perPage
+});
+  
+for await (const response of iterator) {
+    const discussions = response.repository.discussions.nodes
+    console.log(`fetched page of GitHub discussions for: ${owner}/${repo} (${discussions.length})`)
+    for (const discussion of discussions) {
+      discussionsBuffer.push(discussion)
+    }
+}
+
+console.log(`fetched ${discussionsBuffer.length} discussions for: ${owner}/${repo}`)
+
+
+const discussionCategories: any = {};
+const discussionComments = [];
+// now iterate over every discussion object to pull out entries for
+// github_discussion_categories and github_discussion_comments
+for (const discussion of discussionsBuffer) {
+  discussionCategories[discussion.category?.id] = discussion.category;
```

Comment:
I'm not sure if `discussion.category` is ever `null`, but if we're concerned it is, we should probably check for it, or else we'll get into trouble further below when attempting to insert the `null` category:

```suggestion
  if (discussion.category) {
    discussionCategories[discussion.category.id] = discussion.category;
  }
```